frontend/app/controllers/Event.scala

```diff
@@ -59,14 +59,51 @@ class Event(
   private def BuyAction(id: String, onUnauthenticated: RequestHeader => Result) = NoCacheAction andThen recordBuyIntention(id) andThen
     authenticated(onUnauthenticated = onUnauthenticated) andThen subscriptionRefiner(onNonMember = onUnauthenticated)
 
+  def detailsMasterclass(slug: String) = CachedAction { implicit request =>
+    detailsResult(slug)((event: RichEvent) => {
+      if (event.isInstanceOf[MasterclassEvent]) {
+        eventDetail(event)
+      } else {
+        // the url prefix or slug are not canonical so 302 to the correct one
+        Redirect(event.detailsUrl)
+      }
+    });
+  }
+
+  def detailsLive(slug: String) = CachedAction { implicit request =>
+    detailsResult(slug)((event: RichEvent) => {
+      if (event.isInstanceOf[LiveEvent]) {
+        eventDetail(event)
+      } else {
+        // the url prefix or slug are not canonical so 302 to the correct one
+        Redirect(event.detailsUrl)
+      }
+    });
+  }
+
   def details(slug: String) = CachedAction { implicit request =>
+    detailsResult(slug)((event: RichEvent) => {
+      if (event.isInstanceOf[LiveEvent] || event.isInstanceOf[MasterclassEvent]) {
```

Comment:
This is the logic to decide whether to redirect `/event/` to `/masterclass/` or `/live/`.